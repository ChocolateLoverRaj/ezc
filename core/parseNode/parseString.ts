import CoreTokenDatas from '../CoreTokenDatas'
import CoreTokenType from '../CoreTokenType'
import checkToken from './checkToken'
import CoreNodesWithData from './CoreNodesWithData'
import CoreNodeType from './CoreNodeType'
import TryParseNode from './TryParseNode'

const parseString: TryParseNode<CoreNodesWithData[CoreNodeType.STRING]> = async stream => {
  const iterator = stream[Symbol.asyncIterator]()
  const type = {
    enum: CoreNodeType,
    id: CoreNodeType.STRING
  }
  const result = await checkToken(
    iterator,
    type,
    0,
    'Expected string literal',
    { enum: CoreTokenType, id: CoreTokenType.STRING_LITERAL }
  )
  if (!result.success) return result

  return {
    success: true,
    result: {
      node: {
        type,
        data: result.result as CoreTokenDatas[CoreTokenType.STRING_LITERAL]
      },
      length: 1
    }
  }
}

export default parseString
