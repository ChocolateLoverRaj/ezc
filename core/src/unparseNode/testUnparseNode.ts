import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from 'util/dist/arrayToAsyncIterable/arrayToAsyncIterable.js'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: ((await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.result as any).node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
