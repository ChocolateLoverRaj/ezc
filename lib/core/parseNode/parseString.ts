import CoreTokenType from '../CoreTokenType'
import CoreNodesWithData from './CoreNodesWithData'
import CoreNodeType from './CoreNodeType'
import TryParseNode from './TryParseNode'

const parseString: TryParseNode<CoreNodesWithData[CoreNodeType.STRING]> = async stream => {
  const iterator = stream[Symbol.asyncIterator]()
  const { value, done } = await iterator.next()
  if (done === true) return
  if (!(
    value.type.enum === CoreTokenType &&
    value.type.id === CoreTokenType.STRING_LITERAL)) {
    return
  }
  return {
    node: {
      type: {
        enum: CoreNodeType,
        id: CoreNodeType.STRING
      },
      data: value.data as string
    },
    length: 1
  }
}

export default parseString
