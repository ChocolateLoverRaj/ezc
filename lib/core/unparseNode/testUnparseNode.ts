import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from '../arrayToAsyncIterable'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: (await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
