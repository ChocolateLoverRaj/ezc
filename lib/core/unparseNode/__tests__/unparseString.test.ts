import parseString from '../../parseNode/parseString'
import testUnparseNode from '../testUnparseNode'

test('hi', async () => {
  await testUnparseNode(parseString, 'c"hi"')
})
