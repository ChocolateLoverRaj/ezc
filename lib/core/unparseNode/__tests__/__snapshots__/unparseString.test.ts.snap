// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hi 1`] = `"c\\"hi\\""`;
